src/ripple/core/ClosureCounter.h

```diff
@@ -0,0 +1,204 @@
+// who's execution is delayed by a timer or queue.  The reference counting
+// allows a Stoppable to assure that all such postponed closures are
+// completed before the Stoppable declares itself stopped().
+//
+// Ret_t is the type that the counted closure returns.
+// Args_t are the types of the arguments that will be passed to the closure.
+template <typename Ret_t, typename... Args_t>
+class ClosureCounter
+{
+private:
+    std::mutex mutable mutex_ {};
+    std::condition_variable allClosuresDoneCond_ {};  // guard with mutex_
+    bool waitForClosures_ {false};                    // guard with mutex_
+    std::atomic<int> closureCount_ {0};
+
+    // Increment the count.
+    ClosureCounter& operator++()
+    {
+        ++closureCount_;
+        return *this;
+    }
+
+    // Decrement the count.  If we're stopping and the count drops to zero
+    // notify allClosuresDoneCond_.
+    ClosureCounter&  operator--()
+    {
+        // Even though closureCount_ is atomic, we decrement its value under
+        // a lock.  This removes a small timing window that occurs if the
+        // waiting thread is handling a spurious wakeup when closureCount_
+        // drops to zero.
+        std::lock_guard<std::mutex> lock {mutex_};
+
+        // Update closureCount_.  Notify if stopping and closureCount_ == 0.
+        if ((--closureCount_ == 0) && waitForClosures_)
+            allClosuresDoneCond_.notify_all();
+        return *this;
+    }
+
+    // A private template class that helps count the number of closures
+    // in flight.  This allows Stoppables to hold off declaring stopped()
+    // until all their postponed closures are dispatched.
+    template <typename Closure>
+    class CountedClosure
+    {
+    private:
+        ClosureCounter& counter_;
+        std::remove_reference_t<Closure> closure_;
+
+        static_assert (
+            std::is_same<decltype(
+                closure_(std::declval<Args_t>()...)), Ret_t>::value,
+                "Closure arguments don't match ClosureCounter Ret_t or Args_t");
+
+    public:
+        CountedClosure() = delete;
+
+        CountedClosure (CountedClosure const& rhs)
+        : counter_ (rhs.counter_)
+        , closure_ (rhs.closure_)
+        {
+            ++counter_;
+        }
+
+        CountedClosure (CountedClosure&& rhs)
+        : counter_ (rhs.counter_)
+        , closure_ (std::move (rhs.closure_))
+        {
+            ++counter_;
+        }
+
+        CountedClosure (ClosureCounter& counter, Closure&& closure)
+        : counter_ (counter)
+        , closure_ (std::forward<Closure> (closure))
+        {
+            ++counter_;
+        }
+
+        CountedClosure& operator=(CountedClosure const& rhs) = delete;
+        CountedClosure& operator=(CountedClosure&& rhs) = delete;
+
+        ~CountedClosure()
+        {
+            --counter_;
+        }
+
+        Ret_t operator ()(Args_t... args)
+        {
+            return closure_ (args...);
```

Comment:
I think this could be:
```
return closure_(std::forward<Args_t>(args)...);
```
Rationale:  If Args_t[i] is a non-reference type, this will move the args[i] from the by-value parameter list into the closure_ call, potentially avoiding an expensive copy.  If Args_t[i] is a lvalue-reference type, this will pass the args[i] by lvalue reference (const or non-const).  If Args_t[i] is an rvalue-reference type, this will move the args[i] into the closure, which is probably the intent of the client.